### PR TITLE
add istanbul coerage setup

### PR DESCRIPTION
**Pull request by Fa-So**

### CODE REVIEW

**Comment by zemirco** [comment]
Thank you for the PR!

It looks promising. However it creates coverage for both directories, `dist` and `src`. In addition there is the whole `../webpack/...` stuff that shouldn't be in there.

File |   | Statements |   | Branches |   | Functions |   | Lines |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
dist/ua |   | 0% | 0/10 | 100% | 0/0 | 100% | 0/0 | 0% | 0/10
dist/utils |   | 0% | 0/10 | 0% | 0/2 | 0% | 0/1 | 0% | 0/10
dist/webpack: |   | 0% | 0/6 | 0% | 0/7 | 0% | 0/1 | 0% | 0/6
dist/webpack:/node_modules/process |   | 0% | 0/99 | 0% | 0/36 | 0% | 0/16 | 0% | 0/95
dist/webpack:/node_modules/reflect-metadata |   | 0% | 0/489 | 0% | 0/283 | 0% | 0/105 | 0% | 0/465
dist/webpack:/node_modules/webpack/buildin |   | 0% | 0/6 | 0% | 0/4 | 0% | 0/1 | 0% | 0/6
dist/webpack:/src/id |   | 0% | 0/300 | 100% | 0/0 | 100% | 0/0 | 0% | 0/300
dist/webpack:/src/ua |   | 0% | 0/3091 | 0% | 0/2730 | 0% | 0/330 | 0% | 0/3083
dist/webpack:/src/uacp |   | 0% | 0/119 | 0% | 0/64 | 0% | 0/18 | 0% | 0/119
dist/webpack:/src/uasc |   | 0% | 0/212 | 0% | 0/99 | 0% | 0/29 | 0% | 0/212
dist/webpack:/webpack |   | 0% | 0/36 | 0% | 0/26 | 0% | 0/9 | 0% | 0/34
src |   | 0% | 0/27 | 0% | 0/2 | 0% | 0/5 | 0% | 0/27
src/id |   | 100% | 299/299 | 100% | 0/0 | 100% | 0/0 | 100% | 299/299
src/ua |   | 73.69% | 2269/3079 | 21.58% | 1617/7494 | 43.87% | 136/310 | 73.65% | 2264/3074
src/uacp |   | 47.24% | 60/127 | 88.89% | 112/126 | 40.91% | 9/22 | 47.24% | 60/127
src/uasc |   | 31.21% | 54/173 | 51.61% | 80/155 | 52.17% | 12/23 | 31.21% | 54/173
src/utils |   | 100% | 4/4 | 100% | 0/0 | 100% | 1/1 | 100% | 4/4

**Comment by Fa-So** [comment]
In your case, the directories were not deleted

File: tsconfig.json [hunk 1]
```diff
@@ -51,6 +51,7 @@
     // "allowUmdGlobalAccess": true,          /* Allow accessing UMD globals from modules. */
 
     /* Source Map Options */
+    // "sourceMap": true,
```
Comment (zemirco) [inline tsconfig.json]: This is a duplicate of line 13.

File: package.json [hunk 2]
```diff
@@ -4,9 +4,10 @@
   "description": "TypeScript / JavaScript OPC UA client for the browser",
   "main": "index.js",
   "scripts": {
-    "test": "tsc && jest",
-    "tsc": "tsc --skipLibCheck",
-    "watch": "tsc -w --skipLibCheck",
+    "clear": "rm -fr dist coverage",
+    "test": "npm run clear && tsc --sourceMap && jest",
+    "tsc": "npm run clear && tsc --skipLibCheck",
+    "watch": "npm run clear && tsc -w --skipLibCheck",
```
Comment (zemirco) [inline package.json]: I'd like to have two separate scripts. One just for testing and one also for coverage. I want coverage during CI but I don't want coverage every time I run the tests. Especially at the moment where a lot of stuff is changing I just want to see the tests. The coverage puts a lot of noise on the command line output.

File: jest.config.js [hunk 3]
```diff
@@ -115,9 +116,7 @@ module.exports = {
   // rootDir: undefined,
 
   // A list of paths to directories that Jest should use to search for files in
-  // roots: [
-  //   "<rootDir>"
-  // ],
+  roots: ['<rootDir>/dist'],
```
Comment (zemirco) [inline jest.config.js]: I think when we use `roots` we can remove the `testPathIgnorePatterns` config.

**Comment by zemirco** [comment]
Nice, that looks really good. I just had a few comments.

File: jest.config.js (hunk 3 above)
Comment (Fa-So) [reply jest.config.js]: Discovered very well

File: package.json (hunk 2 above)
Comment (Fa-So) [reply package.json]: See if the scripts fit that way

**Pull request merged**